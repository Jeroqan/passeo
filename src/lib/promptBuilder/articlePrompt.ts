/**
 * articlePrompt.ts
 *
 * Ürün bilgilerini alarak SEO uyumlu bir makale oluşturmak için GPT kullanıcı mesajını hazırlar.
 */

export interface ArticlePromptParams {
  productName: string;
  brand: string;
  category: string;
  keyword: string;
  // Opsiyonel olarak eklenebilecek diğer alanlar
  features?: string[];
  tone?: 'Profesyonel' | 'Samimi' | 'Teknik';
}

/**
 * GPT'ye gönderilecek kullanıcı mesajını oluşturur.
 * Prompt, yalnızca HTML çıktısı üretmek üzerine odaklanıyor; başlık ve paragraflarda SEO'ya dair yönergeler içeriyor.
 * @param params Makale için gerekli ürün, marka, kategori ve anahtar kelime bilgileri.
 * @returns Oluşturulan kullanıcı prompt metni.
 */
export function buildArticlePrompt(params: ArticlePromptParams): string {
  const { productName, brand, category, keyword } = params;

  return (
    `Sen profesyonel bir içerik yazarı ve SEO uzmanısın. ` +
    `Kullanıcıya sadece HTML formatında, başlık etiketleri (<h1>, <h2> vb.) ve paragraflar (<p>) kullanarak yanıt ver. ` +
    `Ürünün adı **${productName}**, markası **${brand}**, kategorisi **${category}** ve odak anahtar kelime **${keyword}**. ` +
    `Makale yapısı: önce bir başlık, sonra 3–5 kısa paragraf, her paragrafta anahtar kelimeyi doğal şekilde bir kez kullan. ` +
    `Cümleleri sade ve akıcı tut, gereksiz tekrar yapma.`
  );
}

export function createArticlePrompt({ productName, brand, category, keyword }: {
  productName: string;
  brand: string;
  category: string;
  keyword: string;
}) {
  return `
  Sen profesyonel bir içerik yazarı ve SEO uzmanısın.
  Aşağıdaki bilgileri kullanarak, ${brand} markasının ${productName} ürünü için kategori ${category} ve anahtar kelime "${keyword}" odaklı, SEO uyumlu, akıcı ve bilgilendirici bir makale oluştur. 
  Lütfen metni başlıklara (<h2>, <h3>) ve paragraflara (<p>) ayırarak HTML formatında çıktı ver.
  `;
} 