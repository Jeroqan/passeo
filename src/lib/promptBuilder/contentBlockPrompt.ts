// contentBlockPrompt.ts

// 1. TEKNİK PROMPT
export const createTechnicalPrompt = (
    productName: string,
    brand: string,
    category: string,
    keyword: string
): string => `
Sen profesyonel, nesnel ve sade bir teknik içerik yazarı ve SEO uzmanısın.  
E-ticaret ürünleri için insana özgü, teknik doğruluğu öncelikli ve SEO uyumlu açıklamalar üretirsin.  
Aşağıdaki kurallara ve örnek cümlelere mutlaka uy:

Ürün Bilgileri:
- Ürün: ${productName}
- Marka: ${brand}
- Kategori: ${category}
- Anahtar Kelime: ${keyword}

Kurallar:
- İçerik yalnızca <h2> başlık ve <p> açıklama paragrafından oluşmalı.
- Başlık: 4–6 kelime, teknik ve kısa; ürün adı/marka başlıkta geçmesin.
- Açıklama: Teknik ve sade olsun, ürün adı yalnızca bir kez ve gerekliyse paragrafta kullanılabilir.
- "Bu ürün", "bu model", "özellik sunar/sağlar" gibi kalıplardan kaçın.
- Teknik bir özellik veya değer (DPI, watt, mah, çözünürlük) sadece ürün adı veya verilen bilgilerde net varsa yaz; yoksa "yüksek DPI", "gelişmiş optik sensör" gibi genelleyici ifadeler kullan.  
- Fiyat, kullanıcı yorumu veya reklam ekleme.

🧠 ÇEŞİTLİLİK ve İNSANİLİK:
- Cümlelerin hepsi farklı başlasın, aynı kalıptan kaçın.
- Bazı cümleler kısa bırakılabilir; gereksiz detay verme.
- Bağlaç (ör: "üstelik", "özellikle", "hatta") ve devrik cümle kullanabilirsin.
- Açıklama sonunda veya ortasında gündelik, insana özgü küçük bir ek/dokunuş kullanılabilir.
- Klasik 'sağlar', 'imkan tanır', 'olanak verir' gibi cümleler yerine kısa, vurucu veya gündelik ifadeler tercih et.
- Yazıların aynı kalıpta olması gerekmiyor; gerektiğinde "Çekim hiç bu kadar kolay olmamıştı.", "Bunu herkes ister.", "Detaylarıyla öne çıkıyor." gibi insani eklemeler de ekle.

ÖRNEK AÇIKLAMA CÜMLELERİ:
- "4K video mu? Her detay net."
- "Çekim hiç bu kadar kolay olmamıştı."
- "Düşük ışıkta bile detaylar kaybolmuyor."
- "Kısacası, zorlu çekimlere hazır."
- "Bunu herkes ister."

FORMAT:
<h2>Başlık</h2>
<p>Tek açıklama paragrafı.</p>
`;

// 2. HUMANIZE PROMPT
export const createHumanizePrompt = (htmlContent: string): string => `
Aşağıdaki SEO ürün açıklamasını insan eliyle yazılmış gibi, sade, teknik faydaya odaklı, doğal ve çeşitli bir dille yeniden yaz.

TEMEL KURALLAR:
- Başlık 4–6 kelimeyle yeniden oluşturulmalı (HTML etiketi kullanma).
- Açıklama sadece 2–3 cümle, kısa ve teknik olmalı.
- Ürün adı sadece bir kez geçebilir.
- Tekrarlayan yapı, süslü dil veya örnek/veri tekrarlarından kaçın.

🛡️ ANTI-UYDURMA:
- Teknik değer (DPI, watt, mah, çözünürlük) ürün adı veya verilen bilgilerde net yoksa uydurma değer yazma; gerekirse "yüksek DPI", "uzun pil ömrü" gibi genelleyici terim kullan.

🎯 SEO/KEYWORD:
- Anahtar kelimenin anlamını veya özelliğini anlat, ancak anahtar kelimeyi birebir tekrar etme.

🧠 ÇEŞİTLİLİK:
- Cümleler farklı uzunlukta ve yapıda olsun; bağlaçlı, kısa veya devrik cümleler ekle.
- Aynı cümle yapısı veya paragraf başlangıcı tekrar etmesin.
- Açıklamanın başı ve sonu her seferinde değişebilir.
- Kimi zaman küçük insani gözlem veya gündelik bir dokunuş ekle.

💬 DOĞAL DİL ve İNSANİ DOKUNUŞ:
- "Sağlar", "imkan tanır", "olanak verir" gibi klasik son cümleleri azalt.
- Gerektiğinde kısa vurucu veya gündelik ifadeler kullan:  
  "Gerçekten farkı hissediliyor.", "Kısacası, herkes için ideal.", "Çekim hiç bu kadar kolay olmamıştı.", "Bunu herkes ister."  
- Cümleleri tamamlamak zorunda değilsin; gerektiğinde kısa bırak, gerektiğinde bağlaçla başla.

SEO Ürün Açıklaması:
${htmlContent}
`;

// 3. ZİNCİR FONKSİYONU (async örnek)
export async function generateContentBlock(
    productName: string,
    brand: string,
    category: string,
    keyword: string,
    callGPT: (prompt: string) => Promise<string>
): Promise<string> {
    // 1. Teknik içerik oluştur
    const teknikPrompt = createTechnicalPrompt(productName, brand, category, keyword);
    const teknikCevap = await callGPT(teknikPrompt);

    // 2. Teknik içeriği humanize et
    const humanizePrompt = createHumanizePrompt(teknikCevap);
    const finalContent = await callGPT(humanizePrompt);

    return finalContent; // Sadece son içeriği döndür
} 