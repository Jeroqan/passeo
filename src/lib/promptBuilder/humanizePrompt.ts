export interface HumanizeParams {
  fullText: string;
  selected: string[];
}

// Helper function to escape special characters for regex, moved here for co-location
function escapeRegExp(string: string): string {
  return string.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

function tagSelectedText(fullText: string, selected: string[]): string {
  const paragraphs = fullText.split(/\\n\\s*\\n/);
  const uniqueSentences = [...new Set(selected)];

  const taggedParagraphs = paragraphs.map(paragraph => {
    let tempParagraph = paragraph;
    for (const sentence of uniqueSentences) {
      if (paragraph.includes(sentence)) {
        // Tag the sentence within the paragraph
        tempParagraph = tempParagraph.replace(new RegExp(escapeRegExp(sentence), 'g'), `<rewrite>${sentence}</rewrite>`);
      }
    }
    return tempParagraph;
  });

  return taggedParagraphs.join('\\n\\n');
}


export function buildHumanizePrompt({ fullText, selected }: HumanizeParams): string {
  const taggedText = tagSelectedText(fullText, selected);

  return `
Bir e-ticaret sitesinin teknik ürün içerik editörüsün. Sadece <rewrite> etiketleriyle işaretlenmiş alanları yeniden yazacaksın; cümlelerin yapısını değiştirip daha doğal, teknik ve insan eliyle yazılmış gibi sunacaksın. Blog, inceleme veya hikaye anlatımı havası olmasın; ürünle ilgili teknik ve donanım özelliklerini açık, net ve doğrudan anlatımla yaz. Her cümlede ürün adını tekrarlamaktan kaçın. Cümleleri kelime değiştirerek değil, tamamen farklı yapı ve sıralama ile yaz. Bilgi ve özellikler birebir korunmalı, anlam kaybı olmamalı. KURALLAR: - Yalnızca <rewrite> içindeki cümleleri değiştir. - <rewrite> etiketlerini kaldır. - Paragraf ve başlık düzenini aynen koru. İPUÇLARI: - Her cümle akıcı, profesyonel ve teknik dille yazılsın. - Ürün adını az kullan, yerine gerekli olursa "cihaz", "model", "ürün" gibi ifadeleri tercih et. - Kısa, net ve bilgi odaklı cümleler kullan. - Gereksiz sıfatlardan, kişisel yorumlardan kaçın. - Model adı, teknik terimler varsa olduğu gibi koru. - "Kullanıcı" kelimesini kullanmaktan kaçın. ÖRNEK: Girdi: <rewrite>Bu aksiyon kamerası, yüksek çözünürlüklü 4K videolar çekmenizi sağlar. SJCAM SJ4000 Air modeli, 170 derece geniş açılı lens ile donatılmıştır.</rewrite> Çıktı: Yüksek çözünürlüklü 4K video kaydı imkanı sunar. 170 derece geniş açılı lens ile daha kapsamlı görüntüler elde edilir. Şimdi aşağıdaki metni yukarıdaki kurallara uygun şekilde düzenle: --- Girdi: ${taggedText} --- Çıktı:
    `.trim();
} 