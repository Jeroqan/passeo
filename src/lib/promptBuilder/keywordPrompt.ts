function buildKeywordExtractionPrompt({ productName, brand, category }: { productName: string; brand?: string; category: string; }) {
  const brandLine = brand ? `• Marka: ${brand}` : '';

  return `
Sen ürün özellikleri çıkarmada uzmanlaşmış, nesnel çalışan bir yapay zekâ asistanısın.

Aşağıdaki ürün, gerçek dünyada satışta olan bir üründür.
Lütfen bu ürün için en güncel, teknik özellikleri veya donanım detaylarını yalnızca kısa anahtar kelime şeklinde ve virgülle ayrılmış biçimde listele.

• Ürün Adı: ${productName}
${brandLine}
• Kategori: ${category}

Kurallar:
- Sadece teknik özellikler ve spesifikasyonlar ("hız ayarı", "başlık tipi", "temizleme fırçası" gibi)
- Sayısal değerleri yalnızca doğrudan kullanıcı aramaları için anlamlıysa ("4K video kaydı", "12.000 DPI", "2.0 inç ekran" gibi) dahil edebilirsin. Ancak anlamı bozulmadan mümkünse sayısal olmayan terimlerle ifade etmeye çalış.
- Cümle, açıklama veya reklam dili kullanma
- Marka ismini veya ürün adını tekrar etme
- Yorum yapma (örn: "etkili", "kullanışlı" gibi sıfatlar yok)
- Özellikler güncel veriye dayansın, sadece teorik tahmin olmasın

Çıktı sadece şu formatta olsun:
tüy alma teknolojisi, hız ayarı, başlık tipi, bikini trimmer, ışık sistemi, temizleme fırçası, kablo uzunluğu
  `.trim();
}

export { buildKeywordExtractionPrompt }; 