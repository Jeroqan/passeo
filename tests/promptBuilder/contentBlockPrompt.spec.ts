// tests/promptBuilder/contentBlockPrompt.spec.ts
import { createContentBlockPrompt } from '@/lib/promptBuilder/contentBlockPrompt';

describe('createContentBlockPrompt', () => {
  it('ürün adı, marka, kategori ve anahtar kelimeyi içererek mantıklı bir HTML blok promptu oluşturmalı', () => {
    const productName = 'SuperWidget';
    const brand = 'Acme';
    const category = 'Elektronik';
    const keyword = 'yüksek performans';

    const prompt = createContentBlockPrompt(productName, brand, category, keyword);

    // Temel bilgileri içerdiğini doğrulayalım
    expect(prompt).toContain('SuperWidget');
    expect(prompt).toContain('Acme');
    expect(prompt).toContain('Elektronik');
    expect(prompt).toContain('yüksek performans');

    // Örnek HTML etiketlerinin varlığını kontrol edelim
    expect(prompt).toMatch(/<h2>/);
    expect(prompt).toMatch(/<p>/);
  });

  it('oluşturulan prompt için bir snapshot ile eşleşmeli', () => {
    const prompt = createContentBlockPrompt(
      'Test Product',
      'Test Brand',
      'Test Category',
      'test keyword'
    );
    expect(prompt).toMatchSnapshot();
  });
}); 