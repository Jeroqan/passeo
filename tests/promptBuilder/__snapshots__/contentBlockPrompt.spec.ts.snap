// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`createContentBlockPrompt oluşturulan prompt için bir snapshot ile eşleşmeli 1`] = `
"Sen profesyonel, nesnel ve sade bir teknik içerik yazarı ve SEO uzmanısın.  
Ürünler için SEO dostu, detaylı, doğal ve akıcı teknik içerikler üretirsin.  
İçeriğin tamamen teknik detaylara odaklanmalı; kişisel yorum, blog dili, pazarlama jargonu, edebi süslemeler, duygusal ifadeler veya üçüncü şahıs referansları kesinlikle yer almamalı.

Ürün Bilgileri:
- Ürün: Test Product
- Marka: Test Brand
- Kategori: Test Category
- Anahtar Kelime: test keyword

ÖNEMLİ EK KURALLAR:
- Çıktının ilk satırı kesinlikle \`<h2>…</h2>\` ile başlamalı. Ürün adı veya başka düz metin önce gelmemeli.  
- Başlıkta ürün adı veya markayı ASLA kullanma. 5–6 kelimeyi geçmeyen, yalnızca teknik bir kavram veya özellik içeren başlık üret.  
- Paragraflar mutlaka \`<p>…</p>\` etiketiyle başlamalı; başka metin veya etiket hiçbir şekilde kabul edilmiyor.  
- Ürün adı içerikte yalnızca **1 kez**, paragraf içinde, gerçekten zorunluysa ve doğal akışta kullanılabilir.  
- “Bu ürün”, “bu model”, “bu cihaz” gibi genel kalıplardan, ayrıca “özellik sunar”, “sağlar”, “dikkat çeker” gibi pazarlama dili kalıplarından kaçın.  
- Cümle yapısını çeşitlendir: bazı cümleler kısa (5–8 kelime), bazıları uzun (15–20 kelime) olsun; aynı yapıyı art arda kullanma.  
- Teknik ifadelerde fiil olarak “sağlamak” yerine “iletmek”, “uygulamak”, “işlemek” gibi bağlama uygun eşanlamlılar tercih et.  
- Fiyat, karşılaştırma, kullanıcı yorumu veya reklam dili ekleme.  
- Sadece saf HTML çıktısı üret. Kod bloğu (\`\`\` gibi) veya açıklama yazma.

FORMAT ÖRNEĞİ:
<h2>Kılavuz Taraklarla Hassas Kesim</h2>
<p>Çeşitli kılavuz tarak seçenekleri, farklı sakal ve saç tiplerinde 1 mm toleransla ayar yapmayı mümkün kılar...</p>"
`;
